extensions/consul-config/runtime/src/main/java/io/quarkus/consul/config/runtime/ConsulConfigSourceProvider.java

```diff
@@ -47,27 +48,39 @@ private ConsulConfigSourceProvider(ConsulConfig config, ConsulConfigGateway cons
 
         List<ConfigSource> result = new ArrayList<>(keys.size());
 
+        List<Uni<?>> allUnis = new ArrayList<>();
+
         for (Map.Entry<String, ValueType> entry : keys.entrySet()) {
             String fullKey = config.prefix.isPresent() ? config.prefix.get() + "/" + entry.getKey() : entry.getKey();
-            log.debug("Attempting to look up value of key '" + fullKey + "' from Consul.");
-
-            try {
-                Optional<Response> optionalResponse = consulConfigGateway.getValue(fullKey);
-                if (optionalResponse.isPresent()) {
-                    result.add(
-                            responseConfigSourceUtil.toConfigSource(optionalResponse.get(), entry.getValue(), config.prefix));
-                } else {
-                    String message = "Key '" + fullKey + "' not found in Consul.";
-                    if (config.failOnMissingKey) {
-                        throw new RuntimeException(message);
+            allUnis.add(consulConfigGateway.getValue(fullKey).chain(new Function<Response, Uni<?>>() {
+                @Override
+                public Uni<?> apply(Response response) {
+                    if (response != null) {
+                        result.add(
+                                responseConfigSourceUtil.toConfigSource(response, entry.getValue(),
+                                        config.prefix));
                     } else {
-                        log.info(message);
+                        String message = "Key '" + fullKey + "' not found in Consul.";
+                        if (config.failOnMissingKey) {
+                            return Uni.createFrom().failure(new RuntimeException(message));
+                        } else {
+                            log.info(message);
+                            return Uni.createFrom().nullItem();
+                        }
                     }
+
+                    return Uni.createFrom().nullItem();
                 }
-                log.debug("Done reading value of key '" + fullKey + "'");
-            } catch (IOException e) {
-                throw new UncheckedIOException("An error occurred while attempting to fetch configuration from Consul.", e);
-            }
+            }));
+        }
+
+        try {
+            Uni.combine().all().unis(allUnis).discardItems().await()
```

Comment:
Just to be sure - there is no way we can be on the I/O thread here, right?

Reply:
We actually want to block until these are done because the bootstrap config phase must be completed before any of the runtime init steps